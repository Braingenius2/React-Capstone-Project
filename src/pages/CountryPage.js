import React from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faArrowLeft } from '@fortawesome/free-solid-svg-icons';
import Navbar from '../components/Navbar';
import Hero from '../components/Hero';
import CountryDetail from '../components/CountryDetail';

const CountryPage = () => {
  const { continentName, countryName } = useParams();

  const country = countryName.replace(/-/g, ' ');

  const { countries } = useSelector((state) => state.countries);

  const selectedCountry = countries.find((country) => country.name === country);

  const filteredStats = Object.keys(selectedCountry)
    .filter((item) => (
      item !== 'id'
      && item !== 'name'
      && item !== 'continent'
      && item !== 'population'
      && item !== 'flag'));

  return (
    <div>
      <Navbar
        nav={{ logo: <FontAwesomeIcon icon={faArrowLeft} />, path: `/${continentName}`, pageTitle: 'Country stats' }}
      />
      <main>
        <Hero
          image={selectedCountry?.flag}
          name={country}
          details={[
            { stats: selectedCountry?.population, title: 'inhabitants' },
          ]}
        />
        <section>
          <h2>Country data</h2>
          <ul>
            {
            filteredStats
              .map((title) => (
                <CountryDetail key={title} title={title} amount={selectedCountry[title]} />
              ))
        }
          </ul>
        </section>
      </main>
    </div>
  );
};

export default CountryPage;
