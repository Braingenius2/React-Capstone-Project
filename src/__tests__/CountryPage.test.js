import { Routes, Route, MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { render, screen } from '@testing-library/react';
import continentsReducer, { fetchContinents } from '../redux/continents/continentsSlice';
import { fakeContinentData, fakeCountriesData } from '../mock/apiData';
import countriesReducer, { fetchCountries } from '../redux/countries/countriesSlice';
import CountryPage from '../pages/CountryPage';

describe('The CountryPage component', () => {
  let store;

  beforeEach(async () => {
    const continentState = continentsReducer(undefined, {
      type: fetchContinents.fulfilled,
      payload: fakeContinentData,
    });
    const countriesState = countriesReducer(undefined, {
      type: fetchCountries.fulfilled,
      payload: fakeCountriesData,
    });
    const thunkMiddleware = [thunk];
    const mockStore = configureStore(thunkMiddleware);
    const initialStateContinents = continentState;
    const initialState = countriesState;
    store = mockStore({
      continents: initialStateContinents,
      countries: initialState,
    });
    store.dispatch(getAllContinents());
    store.dispatch(getAllCountries());
  });

  test('should render correctly into the DOM', async () => {
    const { CountryPageComponent } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/Europe/Albania']}>
          <Routes>
            <Route path=":continentName/:countryName" element={<CountryPage />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );
    expect(CountryPageComponent).toMatchSnapshot();
  });

  test('should render the third page of the app with the fake data of Albania country: Name, Cases and deaths', async () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/Europe/Albania']}>
          <Routes>
            <Route path=":continentName/:countryName" element={<CountryPage />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );

    const countryNameAlbania = screen.getByText('Albania');
    const countryCases = screen.getByText('334,726');
    const countryDeaths = screen.getByText('3,602');

    expect(countryNameAlbania).toBeInTheDocument();
    expect(countryCases).toBeInTheDocument();
    expect(countryDeaths).toBeInTheDocument();
  });
});
