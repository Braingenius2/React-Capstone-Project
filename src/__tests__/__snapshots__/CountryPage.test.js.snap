// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The CountryPage component should render correctly into the DOM 1`] = `undefined`;
